core/src/price_estimation/orderbook_based.rs

```diff
@@ -0,0 +1,149 @@
+use super::price_source::PriceSource;
+use crate::models::{BatchId, TokenId};
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use pricegraph::{Orderbook, TokenPair};
+use primitive_types::U256;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+pub struct PricegraphEstimator {
+    orderbook_reader: Arc<dyn StableXOrderBookReading>,
+}
+
+const ONE_OWL: f64 = 1_000_000_000_000_000_000.0;
+
+impl PriceSource for PricegraphEstimator {
+    fn get_prices<'b>(
+        &'b self,
+        tokens: &'b [TokenId],
+    ) -> BoxFuture<'b, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let batch = BatchId::currently_being_solved(SystemTime::now())?;
+            let (account_state, orders) =
+                self.orderbook_reader.get_auction_data(batch.into()).await?;
+            let orderbook = Orderbook::from_elements(orders.iter().map(|order| {
+                order.to_element(U256(
+                    account_state
+                        .read_balance(order.sell_token, order.account_id)
+                        .0,
+                ))
+            }));
+            Ok(self.estimate_prices(tokens, orderbook))
+        }
+        .boxed()
+    }
+}
+
+// Private trait to facilitate testing this module
+#[cfg_attr(test, mockall::automock)]
+trait Pricegraph {
+    fn estimate_price(&self, pair: TokenPair, volume: f64) -> Option<f64>;
+}
+
+impl Pricegraph for Orderbook {
+    fn estimate_price(&self, pair: TokenPair, volume: f64) -> Option<f64> {
+        // Clone the pricegraph so that each estimation is independent of one another.
+        self.clone().fill_market_order(pair, volume)
+    }
+}
+
+impl PricegraphEstimator {
+    #[allow(dead_code)]
+    pub fn new(orderbook_reader: Arc<dyn StableXOrderBookReading>) -> Self {
+        Self { orderbook_reader }
+    }
+
+    fn estimate_prices(
+        &self,
+        tokens: &[TokenId],
+        pricegraph: impl Pricegraph,
+    ) -> HashMap<TokenId, u128> {
+        tokens
+            .iter()
+            .flat_map(|token| {
+                let price_in_token = if token == &TokenId::reference() {
+                    1.0
+                } else {
+                    // Estimate price by selling 1 unit of the reference token for each token.
+                    // We sell rather than buy the reference token because volume is denominated
+                    // in the sell token, for which we know the number of decimals.
+                    let pair = TokenPair {
+                        buy: token.0,
+                        sell: TokenId::reference().0,
+                    };
+                    pricegraph.estimate_price(pair, ONE_OWL)?
+                };
+                let price_in_reference = 1.0 / price_in_token;
```

Comment:
When testing, I noticed that this was backwards before (price directions are hard)

Reply:
I agree :disappointed: